framework/src/modules/interoperability/mainchain/commands/submit_mainchain_cross_chain_update.ts

```diff
@@ -237,19 +184,122 @@ export class SubmitMainchainCrossChainUpdateCommand extends BaseCrossChainUpdate
+		const { ccmID, encodedCCM } = getEncodedCCMAndID(ccm);
+
+		if (!(await this.verifyCCM(context, ccmID))) {
+			return;
+		}
+
+		// create a state snapshot
+		const stateSnapshotID = context.stateStore.createSnapshot();
+		const eventSnapshotID = context.eventQueue.createSnapshot();
+
+		const chainAccountStore = this.stores.get(ChainAccountStore);
+		const beforeCCMForwardingContext: BeforeCCMForwardingContext = { ...context, ccmFailed: false };
+
+		const receivingChainAccount = await chainAccountStore.getOrUndefined(
+			context,
+			ccm.receivingChainID,
+		);
+
+		// do not continue, if
+		// the chain account does not exist
+		// the chain has REGISTERED status
+		if (!receivingChainAccount || receivingChainAccount.status === ChainStatus.REGISTERED) {
+			beforeCCMForwardingContext.ccmFailed = true;
+			if (
+				!(await this._beforeCrossChainMessageForwarding(
+					beforeCCMForwardingContext,
+					eventSnapshotID,
+					stateSnapshotID,
+				))
+			) {
+				return;
+			}
+			await this.bounce(
+				context,
+				encodedCCM.length,
+				CCMStatusCode.CHANNEL_UNAVAILABLE,
+				CCMProcessedCode.CHANNEL_UNAVAILABLE,
+			);
 			return;
 		}
+
+		// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0043.md#sidechain-registration-process
+		// Liveness condition is only applicable to ACTIVE status chains
+		if (
+			!(await this.internalMethod.isLive(context, ccm.receivingChainID, context.header.timestamp))
+		) {
+			beforeCCMForwardingContext.ccmFailed = true;
+
+			/**
+			 * If the receiving chain is active, it means it violated the liveness condition, and we terminate it.
+			 * If the receiving chain is already terminated, terminateChain does nothing.
+			 */
+			await this.internalMethod.terminateChainInternal(context, ccm.receivingChainID);
+
+			/**
+			 * A sidechain terminated message is returned to the sending chain
+			 * to inform them that the receiving chain is terminated.
+			 */
+			await this.internalMethod.sendInternal(
+				context,
+				EMPTY_FEE_ADDRESS,
+				MODULE_NAME_INTEROPERABILITY,
+				CROSS_CHAIN_COMMAND_SIDECHAIN_TERMINATED,
+				ccm.sendingChainID,
+				BigInt(0),
+				CCMStatusCode.OK,
+				codec.encode(sidechainTerminatedCCMParamsSchema, {
+					chainID: ccm.receivingChainID,
+					stateRoot: receivingChainAccount.lastCertificate.stateRoot,
+				}),
+			);
+
+			if (
+				!(await this._beforeCrossChainMessageForwarding(
+					beforeCCMForwardingContext,
+					eventSnapshotID,
+					stateSnapshotID,
+				))
+			) {
+				return;
+			}
+			await this.bounce(
+				context,
+				encodedCCM.length,
+				CCMStatusCode.CHANNEL_UNAVAILABLE,
+				CCMProcessedCode.CHANNEL_UNAVAILABLE,
+			);
+			return;
+		}
+
 		await this.internalMethod.addToOutbox(context, ccm.receivingChainID, ccm);
+
+		// Emit CCM forwarded event.
 		this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
 			code: CCMProcessedCode.SUCCESS,
 			result: CCMProcessedResult.FORWARDED,
```

Comment:
General comment about forward: I don't think that the implementation is equivalent to LIP. 

In particular in [the LIP](https://github.com/LiskHQ/lips/blob/1e27a5ba99a55d71e6d83f44b6e21f18dda98741/proposals/lip-0049.md?plain=1#L391) we have the following pseudocode which run *in all cases* (assuming the first try-except is passed successfully):

```
    # Create a state snapshot.
    baseSnapshot = snapshot of the current state
    try:  
        # Call the beforeCrossChainMessageForwarding functions from other modules.
        # For example, the Token module handles moving the tokens from the escrow.
        # If the ccm forwarding failed,
        # the CCM is bounced back to the sending chain and the fee is assigned to the CCU relayer.
        for each module mdl for which beforeCrossChainMessageForwarding exists:
            mdl.beforeCrossChainMessageForwarding(trs, ccm, ccmFailed)
    except:
        revert state to baseSnapshot
        terminateChain(ccm.sendingChainID)
        emitCCMEvent(ccm, CCM_PROCESSED_RESULT_DISCARDED, CCM_PROCESSED_CODE_INVALID_CCM_BEFORE_CCC_FORWARDING_EXCEPTION)
        return
```

In the implementation, if I understand well, this part is written twice (lines 233-241 and 282-290) which correspond to the two cases where `ccmFailed = True`. So in case `ccmFailed =  False`, i.e., everything is correct, we don't run `beforeCrossChainMessageForwarding` (i.e., don't update escrows) and call `addToOutbox` immediately, which is *not* the intended behavior.

Since a change is needed, I suggest refactoring in order to have the same structure as in the LIP, so that it would also be easier to review and check equivalence.

Reply:
Moreover, in case all unit tests are passing with the current code, it means that we don't have considered all cases. 

In particular the following unit test should exist (and fail using the current specs): If everything is correct, the escrows should be updated (i.e., `beforeCrossChainMessageForwarding` should have been called) AND the ccm should have been added to the outbox of the receiving chain.